lib/events.ts

```diff
@@ -129,7 +129,7 @@ async function sendEvent(balenaUrl: string, event: string, username?: string) {
 		data: Buffer.from(JSON.stringify(trackData)).toString('base64'),
 	};
 	try {
-		await got(url, { searchParams, retry: 0 });
+		await got(url, { searchParams, retry: 0, timeout: 4000 });
```

Comment:
> pdcastro I set the timeout a little bit shorted (4s instead of 5s), let me know what you think.

That's a tricky one. Actually, I had thought of 5s as an intentionally "unreasonably long timeout" so that it would be hopefully clear to anyone, ourselves included, that "something is wrong/broken and needs to be fixed, this is not normal".

For the sake of argument, consider two alternative scenarios:

* The CLI always requires an internet connection of "reasonable quality" (not a 50Kbps dial-up modem!). But in the rare event that the internet is down, some commands (like `balena ssh <ip-address>`) will still work, despite that very noticeable 5s timeout.

* At least for selected commands, the CLI never requires an internet connection. In this case, a 4s or 5s timeout in arguably not acceptable.

These are probably not the only choices, we could go with something in between or some other reasoning. My point is that 4s or 5s timeout is not a good user experience if it was an ordinary behaviour, regular occurrence.

My hope for a solution that addressed #2343 / #2349 / #2372 would be that the 4s or 5s timeout would not be relied on, somehow.

It also depends on the meaning of "no internet connection". In some testing of mine, if I turned WiFi off in the laptop and it had no active network interface at all, that connection timeout would NOT be triggered / relied on. The CLI would actually run faster than usual and exit straight away. That's because, when the process attempts a connection, the OS replies straight away with "sorry, no can do" and the error handling swallows its (as it should), and all is good.

The problem happens when the machine actually has a network connection, but:

1. Some local firewall only allows local communication with servers "on the factory floor" (picture an air gapped manufacturing line, the subject of #2349), and drops network packets to the outside world, like the balena API.

2. The machine has a good, unfiltered internet connection but the balenaCloud backend (API) is down, the subject of #2372.

For the second case, we could argue that relying on a 4s or 5s timeout is OK because it is a very rare occurrence. But for the first case, it would be a regular occurrence and it is harder to say that always having a 4s or 5s timeout is OK. It does not sound OK to me...

But how to solve the first case?

An "obvious" (?) solution is to reduce the timeout, not just from 5s to 4s but reduce it to... hmm... 400ms, or someone's favourite definition of acceptable. :-)

In the old times of Mixpanel, even before Page implemented analytics proxying via the backend (which was then improved to exit "as soon as possible", I think thanks to a suggestion of yours), we had some timeouts in the hundreds of milliseconds that never actually worked properly (for complex implementation reasons related to the Mixpanel module). When we switched to the proxied analytics, we had a brief discussion of what timeout to set and I think it was Page (?) who dared suggest "no timeout at". The reasoning was that we care for analytics and if the balenaCloud backend is not answering fast enough, we need to improve the backend rather than have the CLI drop requests.

An air gapped (filtered internet) manufacturing environment is a different situation to a slow backend, but they share a possible timeout solution.

Long comment, food for thought. If the purpose of this PR is to solve #2372, then:

* Is it really true that a firewall that dropped packets would cause `balena ssh <ip-addr>` to rely on the 4s or 5s timeout?

* If we have to rely on the timeout for #2372, should we reduce it to 400ms or 500ms instead of 4s or 5s?

* Would it make sense to have different timeouts for different CLI commands? E.g. if we classified some commands as "requires internet" or "does not require internet", what about setting the timeouts to 500ms when internet is not required, and 5s when internet is required?
  * Oh, not just "per command" but dynamically, because `balena ssh` only requires an internet connection if the argument is not an IP address or `.local` hostname.

* If a command (as dynamically determined) does not require internet connection, why even attempt analytics and deprecation requests? Answer: because even if an internet connection is not strictly required, "95% of the time" (?!) it would still be available and could be used.

Reply:
Hmm I got confused in my comment, mixing #2349 with #2372. For either issue, a 4s or 5s timeout would be OK if the cause of the timeout is that the balenaCloud backend (API) is down, because this is very rare. However, if the cause of the timeout is an air gapped network (filtered internet connection) which would apply "all of the time", then a 4s of 5s timeout is less clearly acceptable.

I don't know... I am OK with merging this PR as it is. What do you think? :-)

Reply:
Just had another thought / idea for discussion.

I think we are fundamentally saying that `balena ssh <ip-addr>` will _optionally_ use an internet connection (for analytics, Sentry.io and deprecation checks) if one is available, but the difficulty is that we cannot detect whether or not an internet connection is available (quickly without waiting for a timeout that hurts UX) in the scenario of an air gapped network with a firewall that drops packets to/from the outside world.

The CLI understands the BALENARC_UNSUPPORTED=1, BALENARC_NO_SENTRY=1, and BALENARC_NO_ANALYTICS=1 to disable deprecation checks, Sentry.io reporting and analytics reporting respectively. But I think we would **not** want to "advertise" these variables because we could end up with a substantial fraction of users just **always** setting them for all commands. They could add the variables to their `.bashrc` file and make them part of their "standard system setup for the balena CLI"...

However, what about a new variable, say BALENARC_OFFLINE_MODE=1, (or maybe BALENARC_AIR_GAPPED=1) whose behavior would be:

* BALENARC_OFFLINE_MODE=1 tells the CLI never to attempt a connection "to the internet", i.e. balena API, Dockerhub, npm registry etc. Connections to local devices and a local Docker engine would be allowed.

* We would manually label all commands (in some cases, depending on arguments such as `balena ssh <UUID>` _vs._ `balena ssh <ip-addr | host.local>`) regarding whether or not they depend on an internet connection, i.e. whether or not they are compatible with BALENARC_OFFLINE_MODE=1.

* If a user sets BALENARC_OFFLINE_MODE=1, then:
    * If they attempt to run a command that requires an internet connection such as `balena ssh <UUID>`, the CLI would print an error message that that command requires an internet connection. They have to unset the env var.
    * If they attempt to run a command that does NOT require an internet connection such as `balena ssh <ip-addr>`, the CLI simply acts as if BALENARC_UNSUPPORTED=1, BALENARC_NO_SENTRY=1 and BALENARC_NO_ANALYTICS=1 had been set.

* If a user does NOT set BALENARC_OFFLINE_MODE=1, then:
    * If they attempt to run a command that does NOT require an internet connection such as `balena ssh <ip-addr>`, the CLI will attempt submitting analytics and making deprecation checks (and possibly Sentry.io) and the HTTP request may timeout after 5s. We would detect whether the request timed out. If it timed out, we would print a warning message to stderr advising the user, along the lines of _"Timeout submitting analytics event to balenaCloud. If you are using the balena CLI in an air gapped environment without an internet connection, set the BALENARC_OFFLINE_MODE=1 environment variable when using CLI commands that do not strictly require access to balenaCloud, such as `balena ssh <ip-addr>`"_  
I would classify such a warning message as "advertising" BALENARC_OFFLINE_MODE "where needed". Users would not simply always set BALENARC_OFFLINE_MODE=1 because most (?) CLI commands actually do require an internet connection for understandable purposes that users care about, e.g. `balena push`.
    * Otherwise it just works as usual.

Reply:
Thanks for the detailed thoughts. Yeah it's a subtly complex issue.  Yeah I think the solution you've described here with BALENARC_OFFLINE_MODE sounds reasonable.  I'll take a look at the practicalities of implementing that.

Reply:
Indeed! It doesn't look like it would help us in this case though. In the case of a firewall that selectively drops packets (affecting connections to balenaCloud), `lookup` may or may not be blocked, and `connect` (or `secureConnect`) would most likely be blocked. I understand that, if we don't use this object, the 4s timeout we specify applies to the sum of all phases "from lookup to response", which is probably what we want.